### ISSUE TITLE
Feature Request: Ability to export debug information (symbol info, type info)

### ISSUE THREAD

**Issue by LunarLambda** [NationalSecurityAgency/ghidra#2687]
Ghidra can already load and analyze PDB and other debug information, however it cannot currently export it. Line and file information probably isn't feasible, however just having symbol names and maybe type information (structs, function signatures?) would be a huge help for runtime debugging.

**Comment by dragonmacher** [comment]
There is a future where this may be possible.

**Comment by astrelsky** [comment]
Would be great for DWARF in ELF also.

**Comment by johnkeates** [comment]
I was looking around to see if anyone else had this question, and it seems I am indeed not the only one! This is extra useful when you want to run the application in a debugger.

There is a ghidra2dwarf plugin but it doesn't work well on large binaries (and it assumes all disassembled binaries can be turned into complete decompiled code).

**Comment by astrelsky** [comment]
> I was looking around to see if anyone else had this question, and it seems I am indeed not the only one! This is extra useful when you want to run the application in a debugger.
> 
> There is a ghidra2dwarf plugin but it doesn't work well on large binaries (and it assumes all disassembled binaries can be turned into complete decompiled code).

I mean you can use the debugger in ghidra now assuming the one you want to use is supported.

**Comment by johnkeates** [comment]
That's what I thought but it just doesn't seem to work at all (the process gets stuck early on), but running it with Pwndbg (in gdb) with wsym.py and an exported CSV from the Symbol Table from Ghidra and only the addresses and function names does work well enough.

I'm probably just doing something wrong (the connection to gdb locally and over ssh works, it's just that I get no output in Ghidra).

**Comment by astrelsky** [comment]
> That's what I thought but it just doesn't seem to work at all (the process gets stuck early on), but running it with Pwndbg (in gdb) with wsym.py and an exported CSV from the Symbol Table from Ghidra and only the addresses and function names does work well enough.
> 
> I'm probably just doing something wrong (the connection to gdb locally and over ssh works, it's just that I get no output in Ghidra).

Which version are you using? Iirc there was something that sounds similar fixed in 10.0.2.

**Comment by johnkeates** [comment]
I'm at 10.0.1 so if it was a known thing I should probably upgrade  👍

**Comment by astrelsky** [comment]
> I'm at 10.0.1 so if it was a known thing I should probably upgrade 👍

Hopefully it will be fixed. If not open an issue about it. The bystander effect seems to be a bit strong lately.

**Comment by d-millar** [comment]
If 10.0.2 doesn't solve your problem, let us know - we'd like to know if there are still issues!

**Comment by d-millar** [comment]
@astrelsky ”bystander effect”?

**Comment by astrelsky** [comment]
> @astrelsky ”bystander effect”?

Yes. From wikipedia "The bystander effect, or bystander apathy, is a social psychological theory that states that individuals are less likely to offer help to a victim when there are other people present."

Basically "I'm sure someone else has opened an issue about this 'obvious' issue so I won't bother" or something like that.

**Comment by stellarpower** [comment]
This would be very useful from the point of view of being able to use an arbitrary debugger or an IDE the user is more familiar with, or otherwise some need to integrate with another toolchain. I am having difficulties in setting up the Ghidra debugger, and when it does run it seems to be running over breakpoints and I'm not sure why. I expect like many users I'm used to being on linux (where tools are generally installed and just work) and currently stuck reversing something on Windows (where neither is the case). I don't need an awful lot of the tools available - the static analysis is largely enough, then I only need to dip into this application for a couple of lines to see what a variable is and trace whence it came (as the decompiler struggled locating it). So if I could just launch the target application in an IDE I'm already very familiar with, and load in the symbols at startup from file, this would speed up the process greatly. I guess it might also lessen some of the need to fix any issues with the Ghidra debugger so rapidly, if in the meantime there are at least other ways to achieve things.

Many linux distros now support separate installation of debug symbols for system packages, and it seems likely Windows has probably packaged these up separately for some time - so it looks to me that support for debugging in this manner should be pretty good across the board. I've also not ever really delved into what happens underneath with e.g. DWARF, but after many many times of having missynced sources and binaries, if Ghidra has the decompilation, and from that knows the mapping of each instruction to its decompiled source line, I'd hope this is somewhat achievable to implement.
